### PR TITLE
Stats: Add Today/Yesterday toggle support.

### PR DESCRIPTION
**Pull request by timmyc**
One of the most requested features from the old stats system is the ability to toggle between "Today/Yesterday" at the stat component level.  The common use case cited was being able to quickly compare datasets between the two dates, and being able to see yesterday's "Final" numbers.

While the entire set of the prior day's data can be seen by clicking on the chart bar - this toggle is a more _lightweight_ action that only fetches the data for that particular stats module for yesterday, instead of the entire stats page worth of data.

stats_ _the_wordpress_com_blog_ _wordpress_com

Currently, I have hard-coded support for the `day` period only.  Since this is in the context of "Today" and "Yesterday" I'm wondering if we should limit it to when viewing only today's day stat page.  Alternatively, we could change the labels on the segmented control to show the dates so it could be used regardless of which day you are actively viewing.  Extending it to other time periods wouldn't be difficult - again just more of thinking through how to properly label the controls ( "This Month / Last Month" or "Jan 2017 / Dec 2016".

Also the feature is only setup on Post & Pages right now, though it could be turned on for all other redux-powered stats components on the 'Day' page.

__To Test__
- Open a site stats day page
- Interact with the Today / Yesterday control
- Note loading state is triggered when the first fetch of 'Yesterday' happens
- Verify the Summary / View All links also point to the correct day

### CODE REVIEW

File: client/my-sites/stats/stats-module/connected-list.jsx [hunk 1]
```diff
@@ -168,11 +230,25 @@ export default connect( ( state, ownProps ) => {
...
 	const siteSlug = getSiteSlug( state, siteId );
 	const { statType, query } = ownProps;
+	let priorPeriodData;
+	let priorPeriodQuery;
+	let requestingPriorPeriod;
+
+	// Fetch Prior Period
+	if ( ownProps.showPriorPeriodToggle && ! ownProps.summary ) {
+		priorPeriodQuery = merge( {}, query );
+		priorPeriodQuery.date = moment( query.date ).locale( 'en' ).subtract( 1, 'd' ).format( 'YYYY-MM-DD' );
```
Comment (youknowriad) [inline client/my-sites/stats/stats-module/connected-list.jsx]: We could write this something like:

```js
priorPeriodQuery = {
  ...query,
  date: moment( query.date ).locale( 'en' ).subtract( 1, 'd' ).format( 'YYYY-MM-DD' )
}
```

I know, I've read too much of Dan Abramov's code :P

File: client/my-sites/stats/stats-module/connected-list.jsx [hunk 2]
```diff
@@ -119,12 +178,14 @@ class StatsConnectedModule extends Component {
...
 
 		const summaryLink = this.getHref();
-		const displaySummaryLink = data && data.length >= 10;
+		const displaySummaryLink = listData && listData.length >= 10;
 		const isAllTime = this.isAllTimeList();
 
 		return (
 			<div>
 				{ siteId && statType && <QuerySiteStats statType={ statType } siteId={ siteId } query={ query } /> }
+				{ siteId && statType && showPriorPeriodToggle && showingPriorPeriod &&
```
Comment (youknowriad) [inline client/my-sites/stats/stats-module/connected-list.jsx]: Have you considered, showing the query even if the `showingPriorPeriod` is not true? It could act like a "preload" and if we're concerned about having too many requests at the same time, we may delay those (when the currentQueryIsRunning or with a timeout).

**Comment by folletto** [comment]
> One of the most requested features from the old stats system is the ability to toggle between "Today/Yesterday" at the stat component level. 

I'm not sure this is a good idea tho. It increases the complexity of the whole page by a LOT, and I'd rather rework the switch to be sticky (i.e. regardless of where you scroll, there's a top bar that latches to the top with the arrows and the date:

stats-sticky-bar

Btw that's already a pattern in Calypso, we have the component ready: `<StickyPanel>`.

For review: can you point me to the research or feedback for it?

**Comment by timmyc** [comment]
> For review: can you point me to the research or feedback for it?

@folletto - here is the feedback thread: https://en.forums.wordpress.com/topic/stats-feedback?replies=234 and wpcomideas pKdGS-1nv-p2

While re-reading the thread myself, I counted this feature being brought up 7 times.

For comparison here is how it worked in Atlas.  Excuse the styling, those css hit a big regression about a year ago:

so-yesterday
 
> It increases the complexity of the whole page by a LOT

I agree here, to me it adds clutter to the page as well, but a sentiment I _feel_ after reading the feedback is there are users that really **want** the complexity.  And building on top of the flow discussions around porting wp-admin features to calypso - many users had a stats flow that was based on the ability to quickly toggle between today and yesterday at the "module" level.

> switch to be sticky (i.e. regardless of where you scroll, there's a top bar that latches to the top with the arrows and the date)

I like this concept!  The only negative here again is it is a heavier action by loading API requests for the prior period for all stats modules on the page, when the user may only be interested in one particular section.  I was also chatting with @youknowriad about this on Slack, and we do need to consider module-level loading indicators - since data is now persisted in the state tree, when users toggle to a prior period that we have in the state tree, the old data is rendered while the new is being requested.  Related to #10911

Also I'm curious if we could make this toggle less visually obtrusive, maybe allow the toggle via a gridicon in the module header, would that be a good compromise?

**Comment by folletto** [comment]
> a sentiment I feel after reading the feedback is there are users that really want the complexity

This is why feedback has to be evaluated in terms of needs/goals and not specific solutions. It's very rare people want complexity: people want solutions for their problem, but they might not be aware of what that is — and that's fair. It requires introspection. It's our job. :)

---

> While re-reading the thread myself, I counted this feature being brought up 7 times.

I skimmed the 8+1 pages looking for the word "yesterday" and to me it seems something already solved by the two arrows. Even considering "literal" reading of the feedback, I couldn't find any specificity of having them per-component.

The mentions I found were along these lines:

> actually accessable, up-close... directly in front of me

Which I think the sticky bar solves very well at no additional complexity. :) 

I'd also point out that reading the "yesterday" mentions the need is again less in terms of "seeing yesterday and today" but more in "seeing the improvement", like this comment states:

> This allows you to see relatively quickly whether a change or addition you have made to the blog has taken effect

Which has me in total agreement, and I think it's one of the failures of current stats, which is more about raw numbers, requiring the user to "read" it, instead of showing what they want (improvements, spikes, highlights, trends).

As I already mentioned, Stats will require a broader redesign, and that thread is **gold**. The above about today/yesterday is a great example on how the "solution" can differ widely from the "result". The "Day" page should **show more deltas, more increments, more about trends**, and less about hard numbers on today. ;) 

Which is an intuition we started to review with the Follower module in #10347. :) 

---

tl;dr:

* I'd go ahead with the sticky bar for this specific issue
* We REALLY need a holistic redesign of stats

**Comment by timmyc** [comment]
@folletto thanks for the thoughtful response.

> I'd go ahead with the sticky bar for this specific issue

Sounds like a good plan.

> We REALLY need a holistic redesign of stats

💯  When do we start?

**Comment by supernovia** [comment]
> a sentiment I feel after reading the feedback is there are users that really want the complexity

Yeah, the comments from a handful of power users has me gives me the impression that regardless of function, they prefer something visually complex, because visually simple is insulting to them. I don't know whether the average or new user feels that way, but given that no one really _has to do anything to their stats_ to be successful, I suspect some folks might enjoy feeling slightly overwhelmed by the data.

On the flip side, I also noticed at least one active commenter did not realize the bars were clickable until a few days ago, even though I'd mentioned in the original post that they were, and it was an easy way to access past data. So maybe something as a simple as a stats tour would help.

I think if we could offer a few more tables and CSV downloads so power users can do what they'd like with the data (I have heard of some in the past copying into excel, for example) that may help as well.
